contents/java/JavaConcurrency.md

```diff
@@ -0,0 +1,149 @@
+
+### Executors
+While it is easy to create one or two threads and run them, it becomes a problem when your application requires creating 20 or 30 threads for running tasks concurrently. This is where the `Executors` class comes in. It helps you with:
+
+1. **Thread Creation**<br>
+It provides various methods for creating threads, more specifically a pool of threads, uses to run tasks concurrently.
+
+1. **Thread Management**<br>
+It manages the life cycle of the threads in the thread pool. You don’t need to worry about whether the threads in the thread pool are active or busy or dead before submitting a task for execution.
+
+1. **Task submission and execution**<br>
+It provides methods for submitting tasks for execution in the thread pool, and also allows you to decide when the tasks will be executed. For example, a task can be submitted to be executed instantly, scheduled to be executed later or even executed periodically. 
+
+The `Executors` class provides convenient factory methods for creating the `ExecutorService` class. This class manages the lifecycle of tasks in various ways by assigning a `ThreadPool` to it. Different thread pools manage your tasks in different ways along with their own advantages and disadvantages. Some of these include:
+
+* `CachedThreadPool` - Creates new threads as they are needed. This would prove useful for short-lived tasks but otherwise would be resource intensive as it would create too many threads.
+
+* `FixedThreadPool` -  A thread pool with a fixed number of threads. If all threads are active when a new task is submitted, they will wait in queue until a thread becomes available. This can come in handy when you know the exactly how many threads you need, though that may be tricky by itself.
+
+* `ThreadPoolExecutor` - This thread pool implementation adds the ability to configure parameters such as the maximum number of threads it can hold and how long to keep extra threads alive.
+ 
+A simple example of using the `Executors` class is shown below where after passing in the task to be executed, it is automatically managed by the `ExecutorService`. For a more detailed look at the workflow of the `ExecutorService`, see this [in-depth tutorial](https://www.callicoder.com/java-executor-service-and-thread-pool-tutorial/).  
+```java
+ExecutorService executor = Executors.newFixedThreadPool(10); 
+executor.execute(() -> {  
+    System.out.println("Hello " + Thread.currentThread().getName());
+});
+
+```
+
+## Disadvantages of Java Concurrency
+As the saying goes, _there is no free lunch_. While concurrency provides great benefits as mentioned above, it does come with several issues such as:
+
+* **A more complex design**<br>
+Due to the creation of multiple threads and the problem of thread interference, the design becomes more complex as much more work is done to manage these threads such that no problem occurs when the application runs.  
```

Comment:
You can use a concrete example to show how code looks more complex when using concurrency.

Reply:
But then again, wouldn't having complex code be hard to understand for newbies?

Reply:
If you put comments to demarcate each segment, then it wouldn't be :P